x/tokenomics/keeper/settle_pending_claims.go

```diff
@@ -160,25 +161,49 @@ func (k Keeper) SettlePendingClaims(ctx sdk.Context) (
 // This is the height at which the proof window closes.
 // If the proof window closes and a proof IS NOT required -> settle the claim.
 // If the proof window closes and a proof IS required -> only settle it if a proof is available.
-func (k Keeper) getExpiringClaims(ctx sdk.Context) (expiringClaims []prooftypes.Claim) {
+func (k Keeper) getExpiringClaims(ctx sdk.Context) (expiringClaims []prooftypes.Claim, err error) {
 	blockHeight := ctx.BlockHeight()
 
-	// TODO_TECHDEBT: Optimize this by indexing claims appropriately
-	// and only retrieving the claims that need to be settled rather than all
-	// of them and iterating through them one by one.
-	claims := k.proofKeeper.GetAllClaims(ctx)
+	// NB: This error can be safely ignored as on-chain SharedQueryClient implementation cannot return an error.
+	sharedParams, _ := k.sharedQuerier.GetParams(ctx)
+	claimWindowSizeBlocks := sharedParams.GetClaimWindowCloseOffsetBlocks()
+	proofWindowSizeBlocks := sharedParams.GetProofWindowCloseOffsetBlocks()
+
+	previousSessionEndHeight := blockHeight -
+		int64(sharedParams.GetGracePeriodEndOffsetBlocks()+
```

Comment:
Grace period and claim window offset overlap (they should both start counting from session end height). We should omit them from previous session calculation.

Also not that `claimWindowSizeBlocks` and `proofWindowSizeBlocks` could not just added up since they are not necessarily contiguous. We should take their respective offsets into account too.